File: frontend/src/api/internalTransactions.ts
//
// Internal transactions — write-once (create + read only).  To reverse
// a transfer, create a new internal transaction with src and dst
// budgets swapped.
//

import { useAuthStore } from "@/stores/auth";
import { qs } from "@/api/util";
import type { InternalTransaction, InternalTransactionListParams, Paginated } from "@/types/api";

////////////////////////////////////////////////////////////////////////
//
export function listInternalTransactions(
  params?: InternalTransactionListParams,
): Promise<Paginated<InternalTransaction>> {
  return useAuthStore().request<Paginated<InternalTransaction>>(
    `/internal-transactions/${qs(params)}`,
  );
}

////////////////////////////////////////////////////////////////////////
//
export function getInternalTransaction(id: string): Promise<InternalTransaction> {
  return useAuthStore().request<InternalTransaction>(`/internal-transactions/${id}/`);
}

////////////////////////////////////////////////////////////////////////
//
export function createInternalTransaction(
  body: Partial<InternalTransaction>,
): Promise<InternalTransaction> {
  return useAuthStore().request<InternalTransaction>("/internal-transactions/", {
    method: "POST",
    body: body as BodyInit,
  });
}

File: app/moneypools/api/v1/views/internal_transactions.py
"""
DRF viewset for budget-to-budget transfers in the moneypools v1 API.
"""

# 3rd party imports
from django_filters.rest_framework import DjangoFilterBackend
from drf_spectacular.utils import extend_schema, extend_schema_view
from rest_framework import mixins, viewsets
from rest_framework.filters import OrderingFilter
from rest_framework.permissions import IsAuthenticated

# Project imports
from moneypools.models import InternalTransaction
from moneypools.permissions import (
    AccountOwnerCreateMixin,
    AccountOwnerQuerySetMixin,
    IsAccountOwner,
)
from moneypools.service import internal_transaction as internal_transaction_svc

from ..filters import InternalTransactionFilter
from ..serializers.internal_transactions import InternalTransactionSerializer


########################################################################
########################################################################
#
@extend_schema_view(
    list=extend_schema(
        summary="List internal transactions",
        description=(
            "Return budget-to-budget transfers belonging to the "
            "authenticated user's accounts. Filterable by "
            "bank_account, src_budget, dst_budget, and date range "
            "(date_from/date_to). Orderable by created_at."
        ),
    ),
    create=extend_schema(
        summary="Create an internal transaction",
        description=(
            "Transfer money between two budgets in the same bank "
            "account. Required: bank_account (UUID), amount, "
            "src_budget (UUID), and dst_budget (UUID). The "
            "authenticated user is recorded as the actor. Internal "
            "transactions are write-once -- to reverse a transfer, "
            "create a new one with src and dst swapped."
        ),
    ),
    retrieve=extend_schema(
        summary="Get internal transaction details",
        description="Return a single internal transaction by UUID.",
    ),
)
class InternalTransactionViewSet(
    AccountOwnerCreateMixin,
    AccountOwnerQuerySetMixin,
    mixins.CreateModelMixin,
    mixins.RetrieveModelMixin,
    mixins.ListModelMixin,
    viewsets.GenericViewSet,
):
    """Write-once budget-to-budget transfers within a bank account."""

    serializer_class = InternalTransactionSerializer
    queryset = InternalTransaction.objects.select_related(
        "bank_account", "src_budget", "dst_budget", "actor"
    ).all()
    lookup_field = "id"
    permission_classes = [IsAuthenticated, IsAccountOwner]
    filter_backends = [DjangoFilterBackend, OrderingFilter]
    filterset_class = InternalTransactionFilter
    ordering_fields = ["created_at"]
    ordering = ["-created_at"]

    ####################################################################
    #
    def perform_create(self, serializer: InternalTransactionSerializer) -> None:
        """Save the internal transaction with the requesting user as actor."""
        data = serializer.validated_data
        serializer.instance = internal_transaction_svc.create(
            bank_account=data["bank_account"],
            src_budget=data["src_budget"],
            dst_budget=data["dst_budget"],
            amount=data["amount"],
            actor=self.request.user,
            effective_date=data.get("effective_date"),
        )
